### ISSUE TITLE
Using explicit (predefined) validation set for grid search

### ISSUE THREAD

**Issue by PoeteMaudit** [NicolasHug/Surprise#211]
#### Description

I am wondering how can I use a separate validation/test set at the `GridSearchCV `to decide which hyperparameters set is the most suitable for my model.

Specifically, let's suppose that I have a .csv dataset and I simply want to load it and split into one training set and into one validation/test set. Then I want to make `Surprise ``GridSearchCV ` to use the former only for training and the latter only for choosing the best set of hyperparameters.

There are already relevant answered questions on StackOverflow regarding `Scikit-Learn`:
1) https://stackoverflow.com/questions/31948879/using-explict-predefined-validation-set-for-grid-search-with-sklearn
2) https://stackoverflow.com/questions/43764999/python-machine-learning-perform-a-grid-search-on-custom-validation-set?noredirect=1&lq=1
3) https://stackoverflow.com/questions/52579293/gridseachcv-with-separate-training-validation-sets-erroneously-takes-also-into?noredirect=1&lq=1

However, I have the impression that with the `PredefinedKFold` of `Surprise `this is not yet exactly possible while with `sklearn.model_selection.PredefinedSplit` you can simply do it since you create a one-dimensional list with indices which specify which rows of a dataframe  are of the training and which are of the validation set.

**Comment by NicolasHug** [comment]
There's no nice builtin way like in scikit learn but I made a FAQ entry: https://surprise.readthedocs.io/en/stable/FAQ.html#how-to-save-some-data-for-unbiased-accuracy-estimation

**Comment by PoeteMaudit** [comment]
Thank you for your instant reply.

Ok but what you describe at your example/link is not at all what I describe at my post along with the links provided.
I am saying:
1) Split the dataset into set A and set B
2) Compute the weights according to set A
3) Pick the best hyperparameter set according to set B

While you are saying:
1) Split the dataset into set A and set B
2) Compute the weights and pick best hyperparameter set according to set A
3) Simply see how the model performs at set B

Usually the purpose of (cross) validation is "I am saying" that is to pick the best hyperparameters set according to a dataset which is different than the training set.

**Comment by NicolasHug** [comment]
Sorry, I read too fast indeed.

So basically, you want to do grid search but instead of doing CV, you want a predefined training set and a predefined testset (and they both never change)?

I believe this would do the trick:

```python
class ConstantFolds():

    def __init__(self, trainset_size, n_splits=None, random_state=None,
                 shuffle=True):

        self.n_splits = 1
        self.shuffle = shuffle
        self.random_state = random_state
        self.trainset_size = trainset_size

    def split(self, data):

        indices = np.arange(len(data.raw_ratings))

        if self.shuffle:
            get_rng(self.random_state).shuffle(indices)

        raw_trainset = data.raw_ratings[:self.trainset_size]
        raw_testset = data.raw_ratings[self.trainset_size:]

        trainset = data.construct_trainset(raw_trainset)
        testset = data.construct_testset(raw_testset)

        yield trainset, testset

    def get_n_folds(self):

        return self.n_splits
```

I hacked it fast so there might be bugs. If needed it could be extended to look like something like predefinedfolds form sklearn.

It'd be a nice addition BTW, if you want to work on this LMK

**Comment by PoeteMaudit** [comment]
Yes, this is what I am talking about more or less.

(Sorry I am new in Github and I do not even know what LMK is)

**Comment by NicolasHug** [comment]
Let Me Know ^^

**Comment by NicolasHug** [comment]
Can I / you close the issue?
